Add support for BackendV2

rathishcholarajan opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

✅ I have added the tests to cover my changes.
✅ I have updated the documentation accordingly.
✅ I have read the CONTRIBUTING document.
-->

### Summary
Add support for BackendV2


### Details and comments
Copied from https://github.com/Qiskit/qiskit-ibmq-provider/pull/1108
Fixes #238

rathishcholarajan committed:
Construct target and qubit properties

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

rathishcholarajan committed:
Add support for BackendV2

rathishcholarajan committed:
Add release note

rathishcholarajan committed:
Fix black

daka1510 commented:
@rathishcholarajan there are a lot of unrelated formatting changes that also cause the CI to fail. Can you rerun black and commit the changes? Makes it easier to focus on the relevant changes only :). Happy to take a deeper look afterwards.

daka1510 commented on qiskit_ibm_provider/accounts/management.py:
```diff
@@ -95,10 +95,7 @@ def _matching_default(account_name: str) -> bool:
-                kv[0],
-                Account.from_saved_format(kv[1]),
-            ),
+            lambda kv: (kv[0], Account.from_saved_format(kv[1]),),
```
@rathishcholarajan there are a lot of unrelated formatting changes that also cause the CI to fail. Can you rerun black and commit the changes? Makes it easier to focus on the relevant changes only :). Happy to take a deeper look afterwards.

rathishcholarajan replied:
Fixed!

rathishcholarajan committed:
Fix style

rathishcholarajan committed:
Fix docs build

kt474 reviewed (approved):
Everything lgtm - best to wait for @daka1510 to have another pair of eyes review

rathishcholarajan committed:
Merge branch 'main' into backendv2

rathishcholarajan committed:
Merge branch 'main' into backendv2

rathishcholarajan commented:
Let's not merge this until we get an all pass on integration tests in my fork.

daka1510 commented on qiskit_ibm_provider/ibm_backend.py:
```diff
@@ -104,6 +116,77 @@ class IBMBackend(Backend):
     :class:`BackendJobLimit<qiskit_ibm_provider.BackendJobLimit>` instance::
 
         job_limit = backend.job_limit()
+
+    Here is list of attributes available on the ``IBMBackend`` class:
+        * name: backend name.
+        * backend_version: backend version in the form X.Y.Z.
+        * num_qubits: number of qubits.
+        * target: A :class:`qiskit.transpiler.Target` object for the backend.
+        * basis_gates: list of basis gates names on the backend.
+        * gates: list of basis gates on the backend.
+        * local: backend is local or remote.
+        * simulator: backend is a simulator.
+        * conditional: backend supports conditional operations.
+        * open_pulse: backend supports open pulse.
+        * memory: backend supports memory.
+        * max_shots: maximum number of shots supported.
+        * coupling_map (list): The coupling map for the device
+        * supported_instructions (List[str]): Instructions supported by the backend.
+        * dynamic_reprate_enabled (bool): whether delay between programs can be set dynamically
```
What's the source for these attributes? Can we keep it in sync over time?

rathishcholarajan replied:
https://github.com/Qiskit/qiskit-terra/blob/main/qiskit/providers/models/backendconfiguration.py#L253
Yes, these don't change too often.

mtreinish replied:
They're defined in an external schema here: https://github.com/Qiskit/ibm-quantum-schemas/blob/main/schemas/backend_configuration_schema.json which also was originally based on: https://arxiv.org/abs/1809.03452

daka1510 replied:
Thanks for the clarifications. Up to you, @rathishcholarajan if you want to include a reference to the schema and/or original paper here. For users like me it would have been helpful.

daka1510 reviewed (commented):
Looks mostly good, just added 2 questions.

rathishcholarajan commented on qiskit_ibm_provider/utils/backend_converter.py:
```diff
@@ -0,0 +1,142 @@
+)
+
+
+def convert_to_target(
```
There is a copy of this in terra but under test utils which we can't rely upon https://github.com/Qiskit/qiskit-terra/blob/5ed5db90b7f4b5988647aa72384b7833ea1184a7/qiskit/test/mock/utils/backend_converter.py

@mtreinish Do you think this can be moved to terra and consumed by providers? (I don't have a strong opinion either way).

mtreinish replied:
The primary reason I didn't do this directly in terra is I really want to be rid of the `BackendConfiguration`, `BackendProperties`, and `PulseDefaults` objects. They're poorly designed, IBM specific (while terra is supposed to be vendor agnostic), and hard to use. When I added BackendV2 it was in a big part to move away from these objects and I was reluctant to have anything that leverages these backend model objects as part of that.

The second more minor detail is the implementation currently in terra for the fake backends and what was used in the runtime client are subtlety different. The copy in terra's fake backend is closer to what I originally wrote for this converter function and works directly with the parsed json payload dictionaries and goes straight from dict -> Target. While IIRC for the runtime provider @rathishcholarajan adapted it to work with the model objects instead of the raw dictionary.

That being said I think that I agree and we probably will want to add these utility function to terra in the medium term despite my hesitations because I think for terra internally we're going to want to start using the `Target` directly as the sole payload we pass to transpiler passes (instead of as an optional kwarg like it's done today). So having a way for terra to generate a Target from a BackendV1 object (with the model objects) as the first step in `transpile()`. I think for terra 0.21 or 0.22 that would be a very reasonable thing to add, but for the near term this needs to be defined somewhere for the ibm provider package to use.

daka1510 replied:
I guess the challenge is that our current core package (terra) is not supposed to have IBM specific code. But we now have 3 different IBM specific packages with common code requirements. The lack of an IBM specific core package might be the source of some of the discussions of code duplication we had lately.

For this particular case I think we're good as the mid-term plan is to add shared utility functions to terra. Fine with me to live with duplication short-term.

rathishcholarajan committed:
Merge branch 'main' into backendv2

daka1510 reviewed (approved):
LGTM

rathishcholarajan committed:
Merge branch 'main' into backendv2

kt474 committed:
Merge branch 'main' into backendv2

rathishcholarajan committed:
Merge branch 'main' into backendv2

rathishcholarajan commented on qiskit_ibm_provider/utils/backend_converter.py:
```diff
@@ -0,0 +1,142 @@
+    return target
+
+
+def qubit_props_dict_from_props(properties: BackendProperties) -> QubitProperties:
```
@nkanazawa1989 Could you please provide a simple example how it is used today and how V2 breaks it?

V2 still supports the same `backend.properties()` method as V1, so I am not sure how this is a breaking change.

nkanazawa1989 replied:
If I understand correctly the support for `backend.properties()` is just a backward compatibility issue. When I discussed design of `QubitProperties` class with @mtreinish in his old PR, we decided to design the class to be universal for different quantum systems (e.g. two-level system may not have anharmonicity). This is why `QubitProperties` base class has small, or minimum set of properties. Readout assignment error is important for construction of noise model, and anharmonicity is important for evaluation of frequency collision in trasmons. Once we want to use it in the hardware-aware transpiler pass, `QubitProperties` is usually not enough. Currently [`PassManagerConfig`](https://github.com/Qiskit/qiskit-terra/blob/b2f0964f2c43b7f6104db9e28191210cc6a71bd4/qiskit/transpiler/passmanager_config.py#L19-L40) object stores full `backend_properties` and current poor `QubitProperties` doesn't motivate low-level optimization pass developer to use `Target`.

nkanazawa1989 replied:
Note that `Target` class stores qubit properties and passes can use it. This means we don't need to call transpiler passes with `backend_properties` object once we migrate to `Target` model. So we should provide all qubit information that IBM provides as a `QubitProperties` instance.
https://github.com/Qiskit/qiskit-terra/blob/b2f0964f2c43b7f6104db9e28191210cc6a71bd4/qiskit/transpiler/target.py#L256

rathishcholarajan replied:
@nkanazawa1989 Thank you. I hope I understood correctly. What do you think about this https://github.com/Qiskit/qiskit-ibm-provider/pull/317/commits/e83504360a2858cfde0943282711cb1e7621e879 ?

nkanazawa1989 replied:
Thanks I added some comments.

rathishcholarajan committed:
Fix failing test

rathishcholarajan committed:
Fix style

rathishcholarajan committed:
Remove IBMSimulator class to be able to leverage future performance improvements when lazy loading backends for a provider

rathishcholarajan committed:
Merge branch 'main' into backendv2

rathishcholarajan committed:
Merge branch 'main' into backendv2

rathishcholarajan committed:
Fix failing tests

rathishcholarajan committed:
Fix failing test_retrieve_from_retired_backend

rathishcholarajan committed:
Support additional qubit properties available on IBM Quantum backends

nkanazawa1989 commented on qiskit_ibm_provider/utils/backend_converter.py:
```diff
@@ -0,0 +1,150 @@
+            t1=properties.t1(qubit),
+            t2=properties.t2(qubit),
+            frequency=properties.frequency(qubit),
+            anharmonicity=properties.qubit_property(qubit, "anharmonicity"),
```
This value is reported in units of Hz if this convention is not updated with V2 migration. Same for frequency.

rathishcholarajan replied:
Fixed! https://github.com/Qiskit/qiskit-ibm-provider/pull/317/commits/3e725797245fa08423252bc1456fca511b2fcd99

nkanazawa1989 commented on qiskit_ibm_provider/ibm_qubit_properties.py:
```diff
@@ -0,0 +1,68 @@
+        Args:
+            t1: The T1 time for a qubit in us
+            t2: The T2 time for a qubit in us
+            frequency: The frequency of a qubit in GHz
```
See the comment above.

nkanazawa1989 commented on qiskit_ibm_provider/ibm_qubit_properties.py:
```diff
@@ -0,0 +1,68 @@
+        super().__init__(t1=t1, t2=t2, frequency=frequency)
+        self.anharmonicity = anharmonicity
+        self.readout_error = readout_error
+        self.readout_length = readout_length
```
Perhaps `readout_error` and `readout_length` can be removed from here since this is reported by `Target`.

rathishcholarajan replied:
Ok I have just added anharmonicity for now, we can always add more later if need be https://github.com/Qiskit/qiskit-ibm-provider/pull/317/commits/3e725797245fa08423252bc1456fca511b2fcd99

mtreinish replied:
Yeah we can rely on the target for `readout_error` and `readout_length` since that'll be the error and duration of a Measurement operation in the target.

rathishcholarajan committed:
Just add anharmonicity for now, we can always expose more props later if need be now that we have a sub class

Also convert us to s for t1 and t2 and ghz to hz for frequency and anharmonicity

mtreinish commented on qiskit_ibm_provider/utils/backend_converter.py:
```diff
@@ -0,0 +1,155 @@
+    """Uses BackendProperties to construct
+    and return IBMQubitProperties class.
+    """
+    qubit_props = {}
+    for qubit, _ in enumerate(properties.qubits):
+        t_1 = properties.t1(qubit) * 1e-6  # microseconds to seconds
+        t_2 = properties.t2(qubit) * 1e-6  # microseconds to seconds
+        frequency = properties.frequency(qubit) * 1e9  # GHz to Hz
+        try:
+            anharmonicity = (
+                properties.qubit_property(qubit, "anharmonicity")[0] * 1e9
+            )  # GHz to Hz
+        except Exception:  # pylint: disable=broad-except
+            anharmonicity = None
+        qubit_props[qubit] = IBMQubitProperties(  # type: ignore[no-untyped-call]
+            t1=t_1,
+            t2=t_2,
+            frequency=frequency,
+            anharmonicity=anharmonicity,
+        )
+    return qubit_props
```
This should be changed to a list return instead of a dict with int keys.

rathishcholarajan replied:
Oops sorry didn't notice these comments, sure will open a new PR with these changes.

mtreinish commented on qiskit_ibm_provider/utils/backend_converter.py:
```diff
@@ -0,0 +1,155 @@
+        "reset": Reset(),
+    }
+    custom_gates = {}
+    target = Target(num_qubits=configuration.n_qubits)
```
We should call `qubit_props_dict_from_props` once it returns a list and set `qubit_properties=qubit_properties` here from that list return.

mtreinish commented on qiskit_ibm_provider/ibm_backend.py:
```diff
@@ -147,8 +303,72 @@ def _default_options(cls) -> Options:
+        self._convert_to_target()
+        return self._target
+
+    def qubit_properties(
+        self, qubit: Union[int, List[int]]
+    ) -> Union[IBMQubitProperties, List[IBMQubitProperties]]:
+        """Return IBMQubitProperties for a given qubit.
+
+        Args:
+            qubit: The qubit to get the
+                :class:`~qiskit_ibm_provider.IBMQubitProperties` object for. This can
+                be a single integer for 1 qubit or a list of qubits and a list
+                of :class:`~qiskit_ibm_provider.IBMQubitProperties` objects will be
+                returned in the same order
+
+        Returns:
+            IBMQubitProperties or a list of IBMQubitProperties
+        """
+        self._get_properties()
+        if not self._qubit_properties:
+            self._qubit_properties = qubit_props_dict_from_props(self._properties)
+        if isinstance(qubit, int):  # type: ignore[unreachable]
+            return self._qubit_properties.get(qubit)
+        if isinstance(qubit, List):
+            return [self._qubit_properties.get(q) for q in qubit]
+        return None
```
If we set this in the target this can be deleted.
